x-pack/legacy/plugins/advanced_ui_actions/public/custom_time_range_badge.tsx

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { TimeRange } from 'ui/timefilter/time_history';
+import { prettyDuration, commonDurationRanges } from '@elastic/eui';
+
+import { SEARCH_EMBEDDABLE_TYPE } from '../../../../../src/legacy/core_plugins/kibana/public/discover/embeddable/constants';
+import {
+  Action,
+  IEmbeddable,
+  ActionContext,
+  IncompatibleActionError,
+  Embeddable,
+  EmbeddableInput,
+} from '../../../../../src/legacy/core_plugins/embeddable_api/public';
+import { VisualizeEmbeddable } from '../../../../../src/legacy/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable';
+import { VISUALIZE_EMBEDDABLE_TYPE } from '../../../../../src/legacy/core_plugins/kibana/public/visualize/embeddable/constants';
+
+import { CustomizeTimeRangeModal } from './customize_time_range_modal';
+import { doesInheritTimeRange } from './does_inherit_time_range';
+import { OpenModal } from './shim/types';
+import { CommonlyUsedRange } from './types';
+
+export const CUSTOM_TIME_RANGE_BADGE = 'CUSTOM_TIME_RANGE_BADGE';
+
+export interface TimeRangeInput extends EmbeddableInput {
+  timeRange: TimeRange;
+}
+
+function hasTimeRange(
+  embeddable: IEmbeddable | Embeddable<TimeRangeInput>
+): embeddable is Embeddable<TimeRangeInput> {
+  return (embeddable as Embeddable<TimeRangeInput>).getInput().timeRange !== undefined;
+}
+
+function isVisualizeEmbeddable(
+  embeddable: IEmbeddable | VisualizeEmbeddable
+): embeddable is VisualizeEmbeddable {
+  return embeddable.type === VISUALIZE_EMBEDDABLE_TYPE;
+}
+
+export class CustomTimeRangeBadge extends Action {
+  public readonly type = CUSTOM_TIME_RANGE_BADGE;
+  private openModal: OpenModal;
+  private dateFormat: string;
+  private commonlyUsedRanges: CommonlyUsedRange[];
+
+  constructor({
+    openModal,
+    dateFormat,
+    commonlyUsedRanges,
+  }: {
+    openModal: OpenModal;
+    dateFormat: string;
+    commonlyUsedRanges: CommonlyUsedRange[];
+  }) {
+    super(CUSTOM_TIME_RANGE_BADGE);
+    this.order = 7;
+    this.openModal = openModal;
+    this.dateFormat = dateFormat;
+    this.commonlyUsedRanges = commonlyUsedRanges;
+  }
+
+  public getDisplayName({ embeddable }: ActionContext<Embeddable<TimeRangeInput>>) {
+    return prettyDuration(
+      embeddable.getInput().timeRange.from,
+      embeddable.getInput().timeRange.to,
+      commonDurationRanges,
+      this.dateFormat
+    );
+  }
+
+  public getIconType() {
+    return 'calendar';
+  }
+
+  public async isCompatible({ embeddable }: ActionContext) {
+    const isInputControl =
+      isVisualizeEmbeddable(embeddable) &&
+      embeddable.getOutput().visTypeName === 'input_control_vis';
```

Comment:
we shouldn't need this here, as there is no way to add this if vistype is input_control_vis

Reply:
and why do we need doesInheritTimerange check her eand not on custom_time_range_action ?

Reply:
I suppose this and the search ones are unnecessary.  The `doesInheritTimeRange` check is required here because we don't want a panel to show a badge if it is just showing whatever the parent's time range is.  Only if there is a panel override do we want to show the badge.  We don't want that check on `customTimeRangeAction` because we want to let them add a time range override if one doesn't exist - so always show the action in the context menu.